clippy_lints/src/matches/redundant_guards.rs

```diff
@@ -0,0 +1,226 @@
+        {
+            emit_redundant_guards(
+                cx,
+                outer_arm,
+                if_expr.span,
+                pat_binding,
+                field_name,
+                arm.pat.span,
+                arm.guard,
+                &mut app,
+            );
+        }
+        // `Some(x) if let Some(2) = x`
+        else if let Guard::IfLet(let_expr) = guard
+            && let pat = let_expr.pat
+            && let Some((pat_binding, field_name)) = get_pat_binding(cx, let_expr.init, outer_arm)
+        {
+            emit_redundant_guards(
+                cx,
+                outer_arm,
+                let_expr.span,
+                pat_binding,
+                field_name,
+                pat.span,
+                None,
+                &mut app,
+            );
+        }
+        // `Some(x) if x == Some(2)`
+        else if let Guard::If(if_expr) = guard
+            && let ExprKind::Binary(bin_op, local, pat) = if_expr.kind
+            && matches!(bin_op.node, BinOpKind::Eq)
+            && let Some((pat_binding, field_name)) = get_pat_binding(cx, local, outer_arm)
+            && expr_can_be_pat(cx, pat)
+            // Ensure they have the same type. If they don't, we'd need deref coercion which isn't
+            // possible (currently) in a pattern. In some cases, you can use something like
+            // `as_deref` or similar but in general, we shouldn't lint this as it'd create an
+            // extraordinary amount of FPs.
+            //
+            // This isn't necessary in the other two checks, as they must be a pattern already.
+            && cx.typeck_results().expr_ty(local) == cx.typeck_results().expr_ty(pat)
+        {
+            emit_redundant_guards(
+                cx,
+                outer_arm,
+                if_expr.span,
+                pat_binding,
+                field_name,
+                pat.span,
+                None,
+                &mut app,
+            );
+        }
+    }
+}
+
+fn remove_binding(
+    cx: &LateContext<'_>,
+    pat: &Pat<'_>,
+    span: Span,
+    app: &mut Applicability,
+) -> (Cow<'static, str>, Cow<'static, str>) {
+    let pat_start = snippet_with_applicability(cx, pat.span.with_hi(span.lo()), "<start>", app);
+    let pat_end = snippet_with_applicability(cx, pat.span.with_lo(span.hi()), "<end>", app);
+    (pat_start, pat_end)
+}
+
+fn get_pat_binding<'tcx>(
+    cx: &LateContext<'tcx>,
+    guard_expr: &Expr<'_>,
+    outer_arm: &Arm<'tcx>,
+) -> Option<(Span, Option<Ident>)> {
+    if let ExprKind::Path(qpath) = guard_expr.kind
+        && let res = cx.qpath_res(&qpath, guard_expr.hir_id)
+        && let Res::Local(local) = res
+        && !is_local_used(cx, outer_arm.body, local)
+    {
+        let mut is_from_pat = false;
+        outer_arm.pat.each_binding(|_, hir_id, _, _| {
+            if hir_id == local {
+                is_from_pat = true;
+            }
+        });
```

Comment:
Doesn't look like this works, as `hir_id == local` is only true for the first or

Reply:
Ah it's the id of the `Pat` not the one from the `PatKind::Binding`, in that case we can do a `.walk()` instead of `.each_binding()` and check the kind ourselves